### PR TITLE
add SimpleRouter

### PR DESCRIPTION
**Pull request by midnight-commit**

### CODE REVIEW

File: contracts/interfaces/ISimpleRouter.sol [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+import "./../router/interfaces/IYakRouter.sol";
+
+interface ISimpleRouter {
+    error UnsupportedSwap(address _tokenIn, address _tokenOut);
+    error InvalidConfiguration();
+
+    struct SwapConfig {
+        bool useYakSwapRouter;
+        uint8 yakSwapMaxSteps;
+        address oracleAdapter;
```
[the remark this answers is not in the document]
Comment (midnight-commit) [reply contracts/interfaces/ISimpleRouter.sol:13]: This would be for a future extension. We were planning to use oracles for some or all swaps since the query happens on chain. Removing for now.

File: contracts/router/SimpleRouter.sol [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        configureYakSwapDefaults(_yakSwapFallback, _maxStepsFallback, _yakRouter);
+    }
+
+    function configureYakSwapDefaults(bool _yakSwapFallback, uint256 _maxStepsFallback, address _yakRouter) internal {
+        if (address(yakRouter) == address(0) && _yakRouter == address(0)) {
+            revert InvalidConfiguration();
+        }
+        maxStepsFallback = _maxStepsFallback > 0 ? _maxStepsFallback : 1;
+        yakSwapFallback = _yakSwapFallback;
+        yakRouter = yakSwapFallback && _yakRouter > address(0) ? IYakRouter(_yakRouter) : yakRouter;
```
[the remark this answers is not in the document]
Comment (midnight-commit) [reply contracts/router/SimpleRouter.sol:39]: I agree the implementation is a bit too gimmicky. The idea was to allow dev to update other variables easier. I'll dumb down this function, it's hard to read.

File: contracts/router/SimpleRouter.sol [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        view
+        returns (FormattedOffer memory offer)
+    {
+        uint256[] memory amounts = new uint[](_tokens.length);
+        amounts[0] = _amountIn;
+        for (uint256 i; i < _adapters.length; i++) {
+            amounts[i + 1] = IAdapter(_adapters[i]).query(amounts[i], _tokens[i], _tokens[i + 1]);
+        }
+
+        offer = FormattedOffer({amounts: amounts, path: _tokens, adapters: _adapters, gasEstimate: 0});
```
[the remark this answers is not in the document]
Comment (midnight-commit) [reply contracts/router/SimpleRouter.sol:78]: The information is not used by the strategies. I'll keep this like it is for now, save some gas.

File: contracts/strategies/BaseStrategy.sol [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+                    WAVAX.deposit{value: balance}();
+                }
+                if (address(rewardToken) == address(WAVAX)) {
+                    rewardTokenAmount += balance;
+                    continue;
+                }
+            }
+            uint256 amount = IERC20(reward).balanceOf(address(this));
+            if (amount > 0) {
+                FormattedOffer memory offer = simpleRouter.query(amount, reward, address(rewardToken));
```
[the remark this answers is not in the document]
Comment (midnight-commit) [reply contracts/strategies/BaseStrategy.sol:215]: Using yakSwapFallback is a valid SimpleRouter configuration. On a chain with high gas limit and cheap execution we might opt to use YakRouter exclusively and don't configure any "special" routes.

The dependency on SimpleRouter and central management of routes is on purpose. We do want to move away from hardcoded pairs/paths. Also note that the SimpleRouter instance is updatable in BaseStrategy.

If there is a path missing the strategy would malfunction but no rewards and more importantly user funds are at risk. We will test every strategy after deployment and missing configurations will be noticed quickly.

File: contracts/router/SimpleRouter.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (halo3mic) [reply contracts/router/SimpleRouter.sol:78]: I see where you are coming from, makes sense to not incur additional fees for unnecessary functionality. Although, just to put things in perspective, you will be saving 1000 gas on 200k-300k gas call (0.3%-0.5%).

File: contracts/strategies/BaseStrategy.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (halo3mic) [reply contracts/strategies/BaseStrategy.sol:215]: Based on our experiments on L2s I wouldn't count on calling YakSwap queries from transaction. 

On Optimism the gas is incredibly cheap, but there is low gas limit, quickly reached by YakSwap findBestPath even with the more gas friendly configuration. 

On Arbitrum gas limit is higher, but unfortunately gas is much more expensive costing even $5 in tx fees for a simple query. 

Note that these are the results for only the gas-friendly configuration. Now, if users will want to use YakSwap for best price discovery, like on Avalanche now, we will have to enable more complex configurations. These may be costly even on L2s and even more likely won't be executable due to the gas limit. 
Of course, one could deploy their own YakSwap router with the simplest configuration to use as a fallback. But then it is up to you to maintain its list of adapters as liquidity sources change. 

To sum up, I believe it will be difficult to do generalised best path discovery within a transaction. I hope I am wrong, as this would enable many cool solutions.

[the remark this answers is not in the document]
Comment (halo3mic) [reply contracts/strategies/BaseStrategy.sol:215]: Existence of the path between reward and deposit token is essential for the strategy. Thus, it wouldn't hurt to be proactive and check whether it exists during initialisation.

[the remark this answers is not in the document]
Comment (halo3mic) [reply contracts/strategies/BaseStrategy.sol:215]: I now agree with centralised path storing. After all, paths between two tokens could be constantly updated and you would want to update it in inly one place.

[the remark this answers is not in the document]
Comment (midnight-commit) [reply contracts/strategies/BaseStrategy.sol:215]: Was about to add the suggested check but we actually can't. Very often deposit token is not a simple ERC20 but an LP token where there will be no direct path from reward to deposit token. In those cases we override _convertRewardTokenToDepositToken and add liquidity on the underlying protocol.
